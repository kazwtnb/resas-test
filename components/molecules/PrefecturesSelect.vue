<template>
  <UiCheckboxes
    v-model="selectedPrefectures"
    :items="prefectures"
    id-key="prefCode"
    label-key="prefName"
    @add="emit('add', $event)"
    @remove="emit('remove', $event)"
  />
</template>
<script setup lang="ts">
import UiCheckboxes from '~~/components/atoms/Checkbox/UiCheckboxes.vue';
import { Prefecture } from '~~/types/models/resas-model';
import { usePrefectures } from '~~/composables/usePrefectures';

interface Emits {
  (e: 'add', value: Prefecture): void;
  (e: 'remove', value: Prefecture): void;
}
const emit = defineEmits<Emits>();

const { prefectures, selectedPrefectures } = usePrefectures();
</script>
<style lang="scss" scoped></style>
