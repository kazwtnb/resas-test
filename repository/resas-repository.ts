import { Ref } from 'vue';
import { Population, Prefecture } from '~~/types/models/resas-model';

class ResasRepository {
  prefectures(): Ref<Prefecture[]> {
    const { data } = useFetch<Prefecture[]>('/api/resas/prefectures');
    return data;
  }

  async population(pref: Prefecture): Promise<Population | null> {
    const { data } = await useFetch<number[]>('/api/resas/population', {
      params: { prefCode: pref.prefCode }
    });

    return data.value
      ? {
        ...pref,
        data: data.value
      }
      : null;
  }
}

const $resasRepository = new ResasRepository();

export default $resasRepository;
