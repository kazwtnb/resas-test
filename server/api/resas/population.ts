import { PopulationResult, ResasResponse } from '~/types/entity/resas-entity';

export default defineEventHandler(async (event): Promise<number[]> => {
  const prefCode = useQuery(event).prefCode;
  if (!prefCode) {
    return [];
  }

  const $config = useRuntimeConfig();
  const res: ResasResponse<PopulationResult> = await $fetch(
    'api/v1/population/composition/perYear',
    {
      params: {
        prefCode: useQuery(event).prefCode,
        cityCode: '-'
      },
      baseURL: $config.public.RESAS_ENDPOINT_BASE_URL,
      headers: {
        'X-API-KEY': $config.RESAS_API_KEY
      }
    }
  );

  return (
    res?.result?.data
      ?.find((d) => d.label === '総人口')
      ?.data?.map((i) => i.value) || []
  );
});
