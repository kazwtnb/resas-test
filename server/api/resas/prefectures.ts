import { ResasResponse } from '~/types/entity/resas-entity';
import { Prefecture } from '~/types/models/resas-model';

export default defineEventHandler(async (): Promise<Prefecture[]> => {
  const $config = useRuntimeConfig();
  const res: ResasResponse<Prefecture[]> = await $fetch('api/v1/prefectures', {
    baseURL: $config.public.RESAS_ENDPOINT_BASE_URL,
    headers: {
      'X-API-KEY': $config.RESAS_API_KEY
    }
  });
  
  return res?.result;
});
