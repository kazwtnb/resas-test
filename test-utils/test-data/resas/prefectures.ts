import { ResasResponse } from '~~/types/entity/resas-entity';
import { Prefecture } from '~~/types/models/resas-model';

export const testPrefectures = [
  { prefCode: 1, prefName: '北海道' },
  { prefCode: 2, prefName: '青森県' },
  { prefCode: 3, prefName: '岩手県' },
  { prefCode: 4, prefName: '宮城県' },
  { prefCode: 5, prefName: '秋田県' },
  { prefCode: 6, prefName: '山形県' }
];

export const testPrefecturesResponse: ResasResponse<Prefecture[]> = {
  message: null,
  result: testPrefectures
};
