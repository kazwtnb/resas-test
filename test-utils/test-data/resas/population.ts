import { testPrefectures } from '~~/test-utils/test-data/resas/prefectures';
import {
  PopulationDataSet,
  PopulationResult,
  ResasResponse,
  YearlyPopulationItem
} from '~~/types/entity/resas-entity';
import { Population } from '~~/types/models/resas-model';

const testYearlyPopulationItems: YearlyPopulationItem[] = [
  { year: 1960, value: 4206313 },
  { year: 1965, value: 4798653 },
  { year: 1970, value: 5386163 },
  { year: 1975, value: 5923569 },
  { year: 1980, value: 6221638 },
  { year: 1985, value: 6455172 },
  { year: 1990, value: 6690603 }
];

export const testPopulationDataSet: PopulationDataSet[] = [
  { label: '総人口', data: testYearlyPopulationItems },
  { label: '年少人口', data: testYearlyPopulationItems }
];

export const testPopulationResponse: ResasResponse<PopulationResult> = {
  message: null,
  result: {
    boundaryYear: 2015,
    data: testPopulationDataSet
  }
};

export const testPopulation: Population[] = testPrefectures.map((p, index) => ({
  ...p,
  data: [...Array(10)].map((_, i) => i + index)
}));

export const createPopulation = (customize?: Partial<Population>) => {
  return {
    prefCode: 1,
    prefName: '北海道',
    data: [1, 2, 3],
    ...customize
  };
};
