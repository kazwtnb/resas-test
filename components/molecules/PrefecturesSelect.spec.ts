import { beforeEach, describe, expect, test, vi } from 'vitest';
import { ref } from 'vue';
import { mount } from '@vue/test-utils';
import PrefecturesSelect from './PrefecturesSelect.vue';
import { testPrefectures } from '~~/test-utils/test-data/resas/prefectures';
import $resasRepository from '~~/repository/resas-repository';

describe('PrefecturesSelect', () => {
  beforeEach(() => {
    // /serverディレクトリ配下のBFFがバグでテストで動作しないため、Repository層でモック
    // https://github.com/nuxt/framework/issues/3252
    vi.spyOn($resasRepository, 'prefectures').mockReturnValue(
      ref(testPrefectures)
    );
  });

  test('Snapshot', () => {
    const wrapper = mount(PrefecturesSelect);
    expect(wrapper.element).toMatchSnapshot();
  });
});
