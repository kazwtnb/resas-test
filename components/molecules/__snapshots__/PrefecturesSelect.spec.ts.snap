// Vitest Snapshot v1

exports[`PrefecturesSelect > Snapshot 1`] = `
<ul
  class="ui-checkboxes"
  data-v-4ad4a939=""
>
  
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          北海道
        </span>
      </label>
    </div>
  </li>
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          青森県
        </span>
      </label>
    </div>
  </li>
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          岩手県
        </span>
      </label>
    </div>
  </li>
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          宮城県
        </span>
      </label>
    </div>
  </li>
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          秋田県
        </span>
      </label>
    </div>
  </li>
  <li
    data-v-4ad4a939=""
  >
    <div
      class="ui-checkbox"
      data-v-2d44444b=""
      data-v-4ad4a939=""
    >
      <label
        data-v-2d44444b=""
      >
        <input
          data-v-2d44444b=""
          type="checkbox"
        />
        <div
          class="ui-checkbox-box"
          data-v-2d44444b=""
        >
          <i
            class="fa-solid fa-check"
            data-v-2d44444b=""
          />
        </div>
        <span
          data-v-2d44444b=""
        >
          山形県
        </span>
      </label>
    </div>
  </li>
  
</ul>
`;
